src/components/utils/test/posix_thread_test.cc

```diff
@@ -41,287 +43,466 @@ namespace utils_test {
+    // Name must be large enough to keep 16 symbols. Read previous comment
+    char name[kMaxSize];
+    int result = pthread_getname_np(thread->ThreadHandle(), name, sizeof(name));
+    if (!result) {
+      EXPECT_EQ(kTestThreadName, std::string(name));
+    }
+    threadDelegate->WaitFor(test_lock, 10000);
   }
-  cond_var_.WaitFor(test_lock, 10000);
+
   EXPECT_TRUE(threadDelegate->check_value());
-  DeleteThread(thread);
+
   delete threadDelegate;
-  EXPECT_EQ(NULL, thread->delegate());
+  EXPECT_EQ(NULL, thread->GetDelegate());
+  DeleteThread(thread);
 }
 
-TEST(PosixThreadTest, StartThread_ExpectThreadStarted) {
+TEST_F(PosixThreadTest, StartThread_ExpectThreadStarted) {
   // Arrange
   threads::Thread* thread = NULL;
   TestThreadDelegate* threadDelegate = new TestThreadDelegate();
-  AutoLock test_lock(test_mutex_);
-  // Create thread
-  ASSERT_NO_THROW(thread = CreateThread(threadName, threadDelegate));
-  // Start created thread
-  EXPECT_TRUE(
-      thread->start(threads::ThreadOptions(threads::Thread::kMinStackSize)));
-  cond_var_.WaitFor(test_lock, 10000);
+
+  {
+    AutoLock test_lock(*threadDelegate);
+    // Create thread
+    ASSERT_NO_THROW(thread = CreateThread(kThreadName, threadDelegate));
+    // Start created thread
+    EXPECT_TRUE(
+        thread->Start(threads::ThreadOptions(threads::Thread::kMinStackSize)));
+    threadDelegate->WaitFor(test_lock, 10000);
+  }
+
   EXPECT_TRUE(threadDelegate->check_value());
-  DeleteThread(thread);
+
   delete threadDelegate;
-  EXPECT_EQ(NULL, thread->delegate());
+  EXPECT_EQ(NULL, thread->GetDelegate());
+  DeleteThread(thread);
 }
 
-TEST(PosixThreadTest, StartOneThreadTwice_ExpectTheSameThreadStartedTwice) {
+TEST_F(PosixThreadTest, StartOneThreadTwice_ExpectTheSameThreadStartedTwice) {
   // Arrange
   PlatformThreadHandle thread1_id;
   PlatformThreadHandle thread2_id;
   threads::Thread* thread = NULL;
   TestThreadDelegate* threadDelegate = new TestThreadDelegate();
-  AutoLock test_lock(test_mutex_);
-  // Create thread
-  ASSERT_NO_THROW(thread = CreateThread(threadName, threadDelegate));
-  // Start created thread
-  EXPECT_TRUE(
-      thread->start(threads::ThreadOptions(threads::Thread::kMinStackSize)));
-  thread1_id = thread->CurrentId();
-  thread->stop();
-  // Try to start thread again
-  EXPECT_TRUE(
-      thread->start(threads::ThreadOptions(threads::Thread::kMinStackSize)));
-  thread2_id = thread->CurrentId();
-  EXPECT_EQ(thread1_id, thread2_id);
-  cond_var_.WaitFor(test_lock, 10000);
+
+  {
+    AutoLock test_lock(*threadDelegate);
+    // Create thread
+    ASSERT_NO_THROW(thread = CreateThread(kThreadName, threadDelegate));
+    // Start created thread
+    EXPECT_TRUE(
+        thread->Start(threads::ThreadOptions(threads::Thread::kMinStackSize)));
+    thread1_id = thread->CurrentId();
+    threadDelegate->WaitFor(test_lock, 10000);
+    thread->Stop(threads::Thread::kThreadStopTask);
```

Comment:
Add another UT, wihout stop, but with JoinTask

Reply:
@LuxoftAKutsan 
Updated in:https://github.com/LuxoftSDL/sdl_core/pull/50/commits/ca62978ee665d27ecfceffac8caf96ce740ecb51